Gems/Atom/Asset/Shader/Code/Source/Editor/CommonFiles/Preprocessor.cpp

```diff
@@ -391,24 +431,67 @@ namespace AZ
+            for (AZ::IO::Path inShaderScanFolder : scanFoldersVector)
             {
-                shaderScanFolder /= "ShaderLib";
+                auto shaderScanFolder = inShaderScanFolder / "ShaderLib";
                 if (auto it = AZStd::find_if(includePaths.begin(), includePaths.end(), PathCompare(shaderScanFolder));
                     it == includePaths.end())
                 {
                     // the folders constructed this fashion constitute the base of automatic include search paths
                     if (AZ::IO::SystemFile::Exists(shaderScanFolder.c_str()))
                     {
-                        includePaths.emplace_back(AZStd::move(shaderScanFolder.LexicallyNormal().Native()));
+                        if (inShaderScanFolder == projectPath)
+                        {
+                            // use the <project>/ShaderLib - folder only if contents were modified. 
+                            // Otherwise use the DefaultProjectShaderlib folder from the engine
+                            if (FilesModifiedFromTemplate(shaderScanFolder)) {
+                                AZ_TracePrintf(
+                                    "Preprocessing",
+                                    "The files in %s/ShaderLib were modified since initial creation, and have priority over "
+                                    "%s/Gems/Atom/Asset/DefaultProjectShaderlib/. Be aware that "
+                                    "potential upstream changes need to be migrated to the project manually.",
+                                    projectPath.c_str(),
+                                    enginePath.c_str());
+                                useDefaultProjectShaderLibFromEngine = false;
+                                includePaths.emplace_back(AZStd::move(shaderScanFolder.LexicallyNormal().Native()));
+                            }
+                            else {
+                                AZ_TracePrintf(
+                                    "Preprocessing",
+                                    "The files in %s/ShaderLib are unmodified since initial creation: using the versions from "
+                                    "%s/Gems/Atom/Asset/DefaultProjectShaderlib instead. Consider removing the folder from the "
+                                    "project.",
+                                    projectPath.c_str(),
+                                    enginePath.c_str());
+                                continue;
+                            }
+                        }
+                        else
+                        {
+                            includePaths.emplace_back(AZStd::move(shaderScanFolder.LexicallyNormal().Native()));
+                        }
                     }
                 }
             }
 
+            // finally the <engineroot>/Gems/Atom/Asset/DefaultProjectShaderLib if the project doesn't modify it
+
+            if (useDefaultProjectShaderLibFromEngine)
```

Comment:
Is it an either/or situation? I thought the point of having **partial SRGs** in multiple locations was that you could merge them all together into one SRG in the end. So in theory you could have includes in both the DefaultProjectShaderLib and project/ShaderLib working together. Isn't that how it works currently? Are we not including from both engine as well as project?

Reply:
In the previous implementation there only ever was one `scenesrg.srgi`, that had to be located in the project. That file then had to include a file from the engine so the compiler sees all partial definitions eventually. I now simply decide which `scenesrg.srgi` to include, but never both.

To support what you suggest i'd have to create a 'the_real_scenesrg,srgi' (names are hard) in the engine, have that file include the project-specific 'scenesrg.srgi', and change every existing shader (also inside the projects) to include the new file.

But i actually like that approach better, because having to update existing shaders is less of an issue than having to change existing non-code projects: The shaders of existing projects probably have to be updated sometimes anyway, but most users are unlikely to actually write custom shaders.

Reply:
I think its better to keep as it is, pick engine or project srg but never both. This way you don't end up with possibly incompatible things in the end

Reply:
To be honest this is fine for now. My only concern here is that if a project has a custom view/scenesrg.srgi file which was forked from DefaultProjectShaderLib at release X but then the engine version is modified in version X+1 and may require the project's custom view/scenesrg.srgi to be updated. I dont know the original intention of setting up project's view/scene srg but If I had to take a guess it was to allow project developers to extend these two SRGs further for their needs. Anyways I am not in favor of making more changes at this time. If anything, we could consider modifying this system as part of the big material system update which will also probably break older projects.

Reply:
The hash does feel a bit brittle though: i expect this to fail on Linux as is, since the line endings will be different. And if we are unlucky this could even depend only on your git settings, so not even something we can catch with a platform define?

I've added PR #18566 as an alternative solution to this PR now: I prefer the other solution, but I'm happy to go ahead with this one as well.